### PR TITLE
feat(billing.payment.method): handle paypal in context

### PR DESCRIPTION
**Pull request by Jisay**
Use these deps:
- [x] ovh/manager#1455
- [x] ovh/manager#1414

Ref: PAYM-160

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-manager#2544.0 ✘  
* Stage 1
  * Build dependencies and inject Sentry ✔ 
* Stage 2
  * Build manager CA ✔ 
  * Build manager EU ✔ 
  * Build manager US ✔ 
  * Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-manager#2545.0 ✘  
* Stage 1
  * Build dependencies and inject Sentry ✔ 
* Stage 2
  * Build manager CA ✔ 
  * Build manager EU ✔ 
  * Build manager US ✔ 
  * Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-manager#2653.0 ✘  
* Stage 1
  * Build dependencies and inject Sentry ✔ 
* Stage 2
  * Build manager CA ✔ 
  * Build manager EU ✔ 
  * Build manager US ✔ 
  * Tests ✘

File: client/app/billing/payment/method/controller.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  /* =====================================
+  =            INITIALIZATION            =
+  ====================================== */
+
+  $onInit() {
+    // TODO: create a component that handle message from ui router resolve
+    if (this.message) {
+      this.$timeout(() => {
```
Comment (frenautvh) [inline client/app/billing/payment/method/controller.js]: is the timeout necessary ? i hate them ^^

Comment (Jisay) [reply client/app/billing/payment/method/controller.js]: Unfortunatly, yes :-(
If no `$timeout` I don't know what happen, but the Alerter is reseted...

File: client/app/billing/payment/method/add/views/legacyBillingAddress/component.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import controller from './controller';
+import template from './index.html';
+
+export default {
+  name: 'paymentMethodAddLegacyBillingAddressView',
```
Comment (FredericEspiau) [inline client/app/billing/payment/method/add/views/legacyBillingAddress/component.js]: Alphabetical order :)

File: client/app/billing/payment/method/add/views/legacyBankAccount/controller.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    this.isValidIban = ovhPaymentMethodHelper.isValidIban;
+    this.isValidBic = ovhPaymentMethodHelper.isValidBic;
+  }
+
+  $onInit() {
+    this.model.bankAccount = {
+      iban: null,
+      bic: null,
+    };
+    // _.set(this.$state.current, 'sharedModel.legacyBankAccount', this.model);
```
Comment (FredericEspiau) [inline client/app/billing/payment/method/add/views/legacyBankAccount/controller.js]:
```suggestion
```

File: client/app/billing/payment/method/add/views/legacyBankAccount/controller.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+export default class PaymentMethodAddLegacyBankAccountCtrl {
+  /* @ngInject */
+
```
Comment (FredericEspiau) [inline client/app/billing/payment/method/add/views/legacyBankAccount/controller.js]: extra whitespace

File: client/app/billing/payment/method/add/views/legacyBankAccount/controller.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  constructor(ovhPaymentMethodHelper) {
+    // other attributes
```
Comment (FredericEspiau) [inline client/app/billing/payment/method/add/views/legacyBankAccount/controller.js]:
```suggestion
```

File: client/app/billing/payment/method/add/views/legacyBankAccount/component.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import controller from './controller';
+import template from './index.html';
+
+export default {
+  name: 'paymentMethodAddLegacyBankAccountView',
```
Comment (FredericEspiau) [inline client/app/billing/payment/method/add/views/legacyBankAccount/component.js]: Alphabetical order :)

File: client/app/billing/payment/method/component.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import controller from './controller';
+import template from './index.html';
+
+export default {
+  name: 'billingPaymentMethod',
```
Comment (FredericEspiau) [inline client/app/billing/payment/method/component.js]: Alphabetical order :)

File: client/app/billing/payment/method/translations/Messages_fr_FR.json [hunk 8]
```diff
@@ -17,10 +17,12 @@
   "billing_payment_method_table_action_description_edit": "Modifier la description",
   "billing_payment_method_table_action_set_default": "Définir ce moyen de paiement par défaut",
   "billing_payment_method_table_action_delete": "Supprimer ce moyen de paiement",
-  "billing_payment_method_edit_error": "Une erreur est survenue lors de la modification de la description :",
+  "billing_payment_method_edit_error": "Une erreur est survenue lors de la modification de la description. {{ errorMessage }}",
```
Comment (FredericEspiau) [inline client/app/billing/payment/method/translations/Messages_fr_FR.json]:
```suggestion
  "billing_payment_method_edit_error": "Une erreur est survenue lors de la modification de la description: {{ errorMessage }}.",
```

File: client/app/billing/payment/method/translations/Messages_fr_FR.json [hunk 9]
```diff
@@ -17,10 +17,12 @@
...
   "billing_payment_method_edit_success": "La modification de la description de votre moyen de paiement a bien été effectuée.",
-  "billing_payment_method_default_error": "Une erreur est survenue lors de la modification de votre moyen de paiement par défaut.",
+  "billing_payment_method_default_error": "Une erreur est survenue lors de la modification de votre moyen de paiement par défaut. {{ errorMessage }}",
```
Comment (FredericEspiau) [inline client/app/billing/payment/method/translations/Messages_fr_FR.json]:
```suggestion
  "billing_payment_method_default_error": "Une erreur est survenue lors de la modification de votre moyen de paiement par défaut: {{ errorMessage }}.",
```

File: client/app/billing/payment/method/translations/Messages_fr_FR.json [hunk 10]
```diff
@@ -17,10 +17,12 @@
...
   "billing_payment_method_default_success": "Votre moyen de paiement par défaut a bien été modifié.",
-  "billing_payment_method_delete_error": "Une erreur est survenue lors de la suppression du moyen de paiement.",
-  "billing_payment_method_delete_success": "Votre moyen de paiement a bien été supprimé."
-}
\ No newline at end of file
+  "billing_payment_method_delete_error": "Une erreur est survenue lors de la suppression du moyen de paiement. {{ errorMessage }}",
+  "billing_payment_method_delete_success": "Votre moyen de paiement a bien été supprimé.",
+  "billing_payment_method_add_success": "Votre nouveau moyen de paiement a bien été ajouté.",
+  "billing_payment_method_add_error": "Une erreur est survenue lors de l'ajout de votre moyen de paiement. {{ errorMessage }}"
```
Comment (FredericEspiau) [inline client/app/billing/payment/method/translations/Messages_fr_FR.json]:
```suggestion
  "billing_payment_method_add_error": "Une erreur est survenue lors de l'ajout de votre moyen de paiement: {{ errorMessage }}."
```

File: client/app/billing/payment/method/translations/Messages_fr_FR.json [hunk 11]
```diff
@@ -17,10 +17,12 @@
...
+  "billing_payment_method_edit_error": "Une erreur est survenue lors de la modification de la description. {{ errorMessage }}",
   "billing_payment_method_edit_success": "La modification de la description de votre moyen de paiement a bien été effectuée.",
-  "billing_payment_method_default_error": "Une erreur est survenue lors de la modification de votre moyen de paiement par défaut.",
+  "billing_payment_method_default_error": "Une erreur est survenue lors de la modification de votre moyen de paiement par défaut. {{ errorMessage }}",
   "billing_payment_method_default_success": "Votre moyen de paiement par défaut a bien été modifié.",
-  "billing_payment_method_delete_error": "Une erreur est survenue lors de la suppression du moyen de paiement.",
-  "billing_payment_method_delete_success": "Votre moyen de paiement a bien été supprimé."
-}
\ No newline at end of file
+  "billing_payment_method_delete_error": "Une erreur est survenue lors de la suppression du moyen de paiement. {{ errorMessage }}",
```
Comment (FredericEspiau) [inline client/app/billing/payment/method/translations/Messages_fr_FR.json]:
```suggestion
  "billing_payment_method_delete_error": "Une erreur est survenue lors de la suppression du moyen de paiement: {{ errorMessage }}.",
```

File: client/app/billing/payment/method/add/component.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import controller from './controller';
+import template from './index.html';
+
+export default {
+  name: 'billingPaymentMethodAdd',
+  controller,
+  template,
+  bindings: {
```
[the remark this answers is not in the document]
Comment (Jisay) [reply client/app/billing/payment/method/add/component.js]: where is it documented?

File: client/app/billing/payment/method/translations/Messages_fr_FR.json (hunk 8 above)
Comment (Jisay) [reply client/app/billing/payment/method/translations/Messages_fr_FR.json]: Don't know if there is a final dot to the errorMessage or not. So I don't put them in translations ;-)

File: client/app/billing/payment/method/add/index.html [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+                <oui-step-form data-header="{{ :: 'billing_payment_method_add_choose_type' | translate }}"
+                               data-submit-text="{{ $ctrl.isLastStep('paymentMethodType') ? 'common_add' : 'common_next' | translate }}"
+                               data-disabled="$ctrl.loading.paymentMethodTypes"
+                               data-navigation="!$ctrl.isLastStep('paymentMethodType')">
+
+                    <ovh-payment-method-register
+                        data-model="$ctrl.model"
+                        data-registered-payment-methods="$ctrl.paymentMethods"
+                        data-show-set-as-default-choice="true"
+                        data-on-initialized="$ctrl.addSteps.paymentMethodType.loading = false">
```
Comment (marie-j) [inline client/app/billing/payment/method/add/index.html]: Could be extracted to a function maybe

File: client/app/billing/payment/method/add/index.html [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+                        <explanation-text>
```
Comment (marie-j) [inline client/app/billing/payment/method/add/index.html]: Should be prefixed , `explanation-text` is quite vague and could be imported by any library

File: client/app/billing/payment/method/add/index.html [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+                            <div data-ng-if="$ctrl.model.selectedPaymentMethodType">
+                                <div data-ng-if="$ctrl.model.selectedPaymentMethodType.paymentType === $ctrl.OVH_PAYMENT_METHOD_TYPE.CREDIT_CARD">
```
Comment (marie-j) [inline client/app/billing/payment/method/add/index.html]: Maybe a `ng-switch` would be indicated in that case

File: client/app/billing/payment/method/add/index.html [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+                                <p data-ng-if="$ctrl.model.selectedPaymentMethodType.paymentType === $ctrl.OVH_PAYMENT_METHOD_TYPE.PAYPAL"
+                                   data-translate="billing_payment_method_add_selected_paypal_redirect">
+                                </p>
+                            </div>
+                        </explanation-text>
+                    </ovh-payment-method-register>
+
+                    <ovh-payment-method-integration
+                        data-ng-if="$ctrl.isLastStep('paymentMethodType')"
+                        data-on-initialized="$ctrl.onPaymentMethodIntegrationInitialized.bind($ctrl)"
```
Comment (marie-j) [inline client/app/billing/payment/method/add/index.html]: Why do you need to bind ? If you're binding is a `&` it will not be necessary I think

File: client/app/billing/payment/method/add/index.js [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+import 'ovh-ui-angular';
+import '@ovh-ux/ng-translate-async-loader';
+import '@ovh-ux/ng-ovh-payment-method';
+import '@ovh-ux/manager-core';
+
+import routing from './routing';
+import component from './component';
+
+import billingContactViewModule from './views/billingContact';
+import legacyBankAccountViewModule from './views/legacyBankAccount';
```
Comment (marie-j) [inline client/app/billing/payment/method/add/index.js]: Sound weird to have `modules` in `/views` 😄

File: client/app/billing/payment/method/add/views/billingContact/controller.js [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+      .sortContacts(this.contactList);
+    return this.contactList;
+  }
+
+  /* =============================
+  =            Events            =
+  ============================== */
+
+  onExistingContactTabActive() {
+    this.activeTab = 'existing';
```
Comment (marie-j) [inline client/app/billing/payment/method/add/views/billingContact/controller.js]: I'm not sure this is the way tabs should be handled, imo a tab == a state

File: client/app/billing/payment/method/add/views/legacyBillingAddress/controller.js [hunk 19]
```diff
@@ -26,19 +24,8 @@ class LegacyBankAccountModel {
...
-  constructor($state, currentUser) {
-    // dependencies injections
-    this.$state = $state;
-    this.currentUser = currentUser; // from app route resolve
-
-    // other attribute used in views
-    this.model = new LegacyBankAccountModel(currentUser.billingCountry);
-  }
-
+export default class PaymentMethodAddLegacyBillingAddressCtrl {
```
Comment (marie-j) [inline client/app/billing/payment/method/add/views/legacyBillingAddress/controller.js]: Not yours but each concern should have its own file, controller in a `*.controller.js` and model in a `*.class.js`

File: client/app/billing/payment/method/controller.js [hunk 20]
```diff
@@ -0,0 +1,70 @@
+import { ALERTER_ID } from './constants';
+
+export default class BillingPaymentMethodCtrl {
+  /* @ngInject */
+  constructor($timeout, Alerter, OVH_PAYMENT_MEAN_STATUS, OVH_PAYMENT_METHOD_TYPE,
```
Comment (marie-j) [inline client/app/billing/payment/method/controller.js]: Convention is to put the `constants` as the last deps injected

File: client/app/billing/payment/method/edit/controller.js [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+  onPaymentMethodEditFormSubmit() {
+    this.loading.save = true;
+
+    const redirectToParams = {
+      action: 'edit',
+    };
+
+    return this.resolve.onEditFormSubmit(this.model.description)
+      .then(() => this.modalInstance.close(_.merge(redirectToParams, {
```
Comment (marie-j) [inline client/app/billing/payment/method/edit/controller.js]: It would be better if modal had a state

File: client/app/billing/payment/method/index.html [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+           target="_blank" rel="noopener"
+           data-ng-href="{{ $ctrl.guide }}"
+           data-track-on="click"
+           data-track-type="action"
+           data-track-name="payment_types::autoRenew::link::guides_how_to_use_automatic_renewal">
+            <span data-translate="billing_payment_method_autorenew_guide_help_link"></span>
+            <span class="oui-icon oui-icon-external_link"
+                  aria-hidden="true">
+            </span>
+            <span class="sr-only" data-translate="common_new_tab"></span>
```
Comment (marie-j) [inline client/app/billing/payment/method/index.html]: 👍 (even though the translation should be brought by the module itself)

File: client/app/billing/payment/method/index.html [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+                    data-type="options"
+                    data-type-options="$ctrl.tableFilterOptions.status"
+                    data-sortable
+                    data-filterable>
+            <span data-ng-bind="$ctrl.ovhPaymentMethodHelper.getPaymentMethodStatusText($row)"></span>
+        </oui-column>
+
+        <oui-action-menu data-align="end"
+                         data-compact>
+            <oui-action-menu-item data-text="{{ :: 'billing_payment_method_table_action_download_procedure' | translate }}"
```
Comment (marie-j) [inline client/app/billing/payment/method/index.html]: I think `text` is a deprecated attribute

File: webpack.config.js [hunk 24]
```diff
@@ -84,5 +84,8 @@ module.exports = (env = {}) => {
         __WEBPACK_REGION__: `'${env.region.toUpperCase()}'`,
       }),
     ],
+    optimization: {
+      runtimeChunk: 'single',
```
Comment (marie-j) [inline webpack.config.js]: Maybe this should be extracted to another PR

File: client/app/billing/payment/method/add/index.html (hunk 14 above)
Comment (FredericEspiau) [reply client/app/billing/payment/method/add/index.html]: Let's prefix all components with at least `ovh` :)

Comment (Jisay) [reply client/app/billing/payment/method/add/index.html]: This is not a component but a transclusion inside the component

File: client/app/billing/payment/method/add/index.html (hunk 15 above)
Comment (Jisay) [reply client/app/billing/payment/method/add/index.html]: yes it could. But with `ng-switch` it's not possible to use the enum. What is the best?

File: client/app/billing/payment/method/add/index.html (hunk 16 above)
Comment (Jisay) [reply client/app/billing/payment/method/add/index.html]: Without I lose the reference of my controller in my callback

File: client/app/billing/payment/method/controller.js (hunk 20 above)
Comment (Jisay) [reply client/app/billing/payment/method/controller.js]: :-)
I'm a little bit lost. In some other comment, we need to order injection in alphabetical way and here we need to put `constants` at the end :'(

File: client/app/billing/payment/method/add/index.js (hunk 17 above)
Comment (Jisay) [reply client/app/billing/payment/method/add/index.js]: I wasn't sure about how to do because I didn't know how to handle the translations of specific views.
During run of this module maybe?

File: client/app/billing/payment/method/index.html (hunk 23 above)
Comment (Jisay) [reply client/app/billing/payment/method/index.html]: done

File: client/app/billing/payment/method/add/index.html (hunk 15 above)
Comment (Jisay) [reply client/app/billing/payment/method/add/index.html]: done

File: client/app/billing/payment/method/add/index.html (hunk 14 above)
Comment (Jisay) [reply client/app/billing/payment/method/add/index.html]: done

File: client/app/billing/payment/method/controller.js (hunk 20 above)
Comment (marie-j) [reply client/app/billing/payment/method/controller.js]: Angular deps first, then  services then constants. (but for each type you keep alphabetical order) 
Let's say I have `constructor($translate, MY_CONSTANTS, Alerter, $q, MyService`)
it will become `constructor($q, $translate, Alerter, MyService, MY_CONSTANTS)`
Hope it makes more sense now 😉

File: client/app/billing/payment/method/add/index.js (hunk 17 above)
Comment (marie-j) [reply client/app/billing/payment/method/add/index.js]: I meant the folder name is ambiguous, and translations should be specific to module and you should be able to run your module without its parent

**Comment by ovh-cds** [comment]
CDS Report deploy_uxci_2#3001.0 ✘  
* deploy_uxci
  * deploy_uxci ✘

**Comment by ovh-cds** [comment]
CDS Report build-manager#3001.2 ■   
* Stage 1
  * Build dependencies and inject Sentry ■

File: client/app/billing/payment/method/add/component.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (FredericEspiau) [reply client/app/billing/payment/method/add/component.js]: It's not so do whatever you want I guess 😄
